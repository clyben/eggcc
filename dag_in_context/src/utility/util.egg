(function ListExpr-length (ListExpr) i64)
(function ListExpr-ith (ListExpr i64) Expr :unextractable)
(function ListExpr-suffix (ListExpr i64) ListExpr :unextractable)
(function Append (ListExpr Expr) ListExpr :unextractable)

(rule ((Switch pred inputs branch)) ((union (ListExpr-suffix branch 0) branch)) :ruleset always-run)

(rule ((= (ListExpr-suffix top n) (Cons hd tl)))
    ((union (ListExpr-ith top n) hd)
     (union (ListExpr-suffix top (+ n 1)) tl)) :ruleset always-run)

(rule ((= (ListExpr-suffix list n) (Nil)))
    ((set (ListExpr-length list) n)) :ruleset always-run)

(rewrite (Append (Cons a b) e)
   (Cons a (Append b e))
   :ruleset always-run)
(rewrite (Append (Nil) e)
   (Cons e (Nil))
   :ruleset always-run)

(function tuple-length (Expr) i64 :unextractable)

(rule ((HasType expr (TupleT tl))
       (= len (TypeList-length tl)))
      ((set (tuple-length expr) len)) :ruleset always-run)

;; Create a Get for every index, and rewrite it to see through Concat
(rule ((Single expr)) ((union (Get (Single expr) 0) expr)) :ruleset always-run)
;; initial get
(rule ((> (tuple-length tuple) 0))
      ((Get tuple 0))
      :ruleset always-run)
;; next get
(rule ((= len (tuple-length tuple))
       (= ith (Get tuple i))
       (< (+ i 1) len)
       )
       ((Get tuple (+ 1 i)))
       :ruleset always-run)

;; descend left
(rule ((Get (Concat expr1 expr2) i)
       (= (tuple-length expr1) len1)
       (< i len1))
      ((union (Get (Concat expr1 expr2) i)
              (Get expr1 i)))
       :ruleset always-run)
;; descend right
(rule ((Get (Concat expr1 expr2) i)
       (= (tuple-length expr1) len1)
       (>= i len1))
      ((union (Get (Concat expr1 expr2) i)
              (Get expr2 (- i len1))))
       :ruleset always-run)


(relation IsLeaf (Expr))
(rule ((Arg ty)) ((IsLeaf (Arg ty))) :ruleset always-run)
(rule ((Const val ty)) ((IsLeaf (Const val ty))) :ruleset always-run)
(rule ((Empty ty)) ((IsLeaf (Empty ty))) :ruleset always-run)

(rewrite (Concat (Concat a b) c)
         (Concat a (Concat b c))
         :ruleset always-run)



;; This Expr size actually do the same thing as cost function.
(function Expr-size (Expr) i64 :unextractable :merge (min old new) )
(function ListExpr-size (ListExpr) i64 :unextractable :merge (min old new))

(rule ((= expr (Function name tyin tyout out)) 
       (= sum (Expr-size out))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Const n ty))) 
      ((set (Expr-size expr) 1))  :ruleset always-run)

(rule ((= expr (Bop op x y)) 
       (= sum (+ (Expr-size y) (Expr-size x)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Uop op x)) 
       (= sum (Expr-size x))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Get tup i)) 
       (= sum (Expr-size tup))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Concat x y)) 
       (= sum (+ (Expr-size y) (Expr-size x)))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Single x)) 
       (= sum (Expr-size x))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)

(rule ((= expr (Switch pred branches)) 
       (= sum (+ (ListExpr-size branches) (Expr-size pred)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (If pred then else)) 
       (= sum (+ (Expr-size else) (+ (Expr-size then) (Expr-size pred))))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (DoWhile in pred-and-output)) 
       (= sum (+ (Expr-size pred-and-output) (Expr-size in)))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (Arg ty))) 
      ((set (Expr-size expr) 1)) :ruleset always-run)

(rule ((= expr (Call func arg)) 
       (= sum (Expr-size arg))) 
      ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((Empty ty)) ((set (Expr-size (Empty ty)) 0))  :ruleset always-run)

(rule ((= expr (Cons hd tl)) 
       (= sum (+ (ListExpr-size tl) (Expr-size hd)))) 
      ((set (ListExpr-size expr) sum)) :ruleset always-run)

(rule ((Nil)) 
      ((set (ListExpr-size (Nil)) 0))  :ruleset always-run)

(rule ((= expr (Alloc id e state ty)) ;; do state edge's expr should be counted?
        (= sum (Expr-size e))) 
        ((set (Expr-size expr) (+ sum 1))) :ruleset always-run)

(rule ((= expr (InContext assumption e)) 
       (= sum (Expr-size e))) 
      ((set (Expr-size expr) sum)) :ruleset always-run)


;;                tuple start len
(function SubTuple (Expr i64 i64) Expr :unextractable)
(rewrite (SubTuple expr x 1)
         (Single (Get expr x)) :ruleset always-run)
(rule ((= f (SubTuple expr a b)) (> b 1))
      ((union f 
         (Concat (Single (Get expr a))
                 (SubTuple expr (+ a 1) (- b 1))))) :ruleset always-run)
