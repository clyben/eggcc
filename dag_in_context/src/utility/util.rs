#[cfg(test)]
use crate::{egglog_test, interpreter::Value};

#[test]
fn test_list_util() -> crate::Result {
    use crate::ast::emptyt;
    let emptyt = emptyt();
    let build = format!(
        "
		(let list (Cons (Const (Int 0) {emptyt} (NoContext))
                  (Cons (Const (Int 1) {emptyt} (NoContext))
                  (Cons (Const (Int 2) {emptyt} (NoContext))
                  (Cons (Const (Int 3) {emptyt} (NoContext))
                  (Cons (Const (Int 4) {emptyt} (NoContext)) (Nil)))))))
		(let expr (Switch (Const (Int 1) {emptyt} (NoContext)) (Empty {emptyt} (NoContext)) list))
	"
    );
    let check = format!(
        "
		(check (= (ListExpr-ith list 1) (Const (Int 1) {emptyt} (NoContext))))
        (check (= (ListExpr-ith list 4) (Const (Int 4) {emptyt} (NoContext))))
        (check (= (ListExpr-length list) 5))
	"
    );
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn append_test() -> crate::Result {
    use crate::ast::emptyt;
    let emptyt = emptyt();
    let build = format!(
        "
        (let appended
            (Append
                (Cons (Const (Int 0) {emptyt} (NoContext)) (Cons (Const (Int 1) {emptyt} (NoContext)) (Nil)))
                (Const (Int 2) {emptyt} (NoContext))))
    "
    );

    let check = format!("
        (check (
            =
            (Cons (Const (Int 0) {emptyt} (NoContext)) (Cons (Const (Int 1) {emptyt} (NoContext)) (Cons (Const (Int 2) {emptyt} (NoContext)) (Nil))))
            appended
        ))
    ");
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn test_tuple_ith() -> crate::Result {
    use crate::ast::emptyt;
    let emptyt = emptyt();
    let build = format!(
        "
    (let tup (Concat
                  (Concat (Single (Const (Int 0) {emptyt} (NoContext))) (Single (Const (Int 1) {emptyt} (NoContext))))
                  (Concat (Single (Const (Int 2) {emptyt} (NoContext))) (Single (Const (Int 3) {emptyt} (NoContext))))))
    
    ;; with print
    (let tup2 (Concat 
                (Concat 
                    (Single (Bop (Print) (Const (Int 0) {emptyt} (NoContext)) (Arg (Base (StateT)) (NoContext))))
                    (Concat (Single (Const (Int 1) {emptyt} (NoContext))) 
                                (Single (Const (Int 2) {emptyt} (NoContext)))))
                (Single (Const (Int 3) {emptyt} (NoContext)))))
    "
    );

    let check = format!(
        "
    (check (= (Get tup 0) (Const (Int 0) {emptyt} (NoContext))))
    (check (= (Get tup 1) (Const (Int 1) {emptyt} (NoContext))))
    (check (= (Get tup 2) (Const (Int 2) {emptyt} (NoContext))))
    (check (= (Get tup 3) (Const (Int 3) {emptyt} (NoContext))))
    (check (= 4 (tuple-length tup)))
    (fail (check (Get tup 4)))

    (check (= (Get tup2 0) (Bop (Print) (Const (Int 0) {emptyt} (NoContext)) (Arg (Base (StateT)) (NoContext)))))
    (check (= (Get tup2 1) (Const (Int 1) {emptyt} (NoContext))))
    (check (= (Get tup2 2) (Const (Int 2) {emptyt} (NoContext))))
    (check (= (Get tup2 3) (Const (Int 3) {emptyt} (NoContext))))
    (check (= 4 (tuple-length tup2)))
    (fail (check (Get tup2 4)))
    "
    );
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn test_expr_size() -> crate::Result {
    use crate::ast::*;
    let pureloop = dowhile(
        single(int(1)),
        parallel!(
            less_than(getat(0), int(3)),
            get(switch!(int(2); parallel!(int(4), int(5))), 0)
        ),
    )
    .with_arg_types(emptyt(), tuplet!(intt()));
    let build: String = format!("(let loop {})", pureloop);

    let check = "(check (= 9 (Expr-size loop)))";
    egglog_test(
        build.as_str(),
        check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn test_sub_tuple() -> crate::Result {
    use crate::ast::*;
    let ty = tuplet!(intt(), intt(), intt());
    let arg = arg().with_arg_types(ty.clone(), ty.clone());
    let build = format!("(let expr (SubTuple {} 0 3))", arg);
    let out = concat(
        single(getat(0)),
        concat(single(getat(1)), single(getat(2))),
    )
    .with_arg_types(ty.clone(), ty);

    let check = format!("(check (= expr {}))", out);
    egglog_test(
        build.as_str(),
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}
