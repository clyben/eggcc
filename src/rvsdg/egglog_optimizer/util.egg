;; ####################################
;; implementation of PassThroughArguments
;; Creates a vec of arguments
;; (vec-of (Arg 0) (Arg 1) ...) with length i
(function PassThroughArguments (i64) VecOperand :unextractable)



;; (how many arguments to generate, vector so far)
(function PassThroughArgumentsHelper (i64 VecOperand) VecOperand :unextractable)

(rewrite (PassThroughArguments i)
  (PassThroughArgumentsHelper i (VO (vec-of)))
  :ruleset subst)

(rule ((= lhs (PassThroughArgumentsHelper i (VO rest)))
       (< (vec-length rest) i))
      ((union lhs
        (PassThroughArgumentsHelper i
            (VO (vec-push rest (Arg (vec-length rest)))))))
      :ruleset subst)

(rule ((= lhs (PassThroughArgumentsHelper i (VO rest)))
       (= (vec-length rest) i))
      ((union lhs (VO rest)))
      :ruleset subst)



;; Project each argument out of a body
(function BodyToVecOperand (i64 Body) VecOperand :unextractable)
;; current index, body length, body, and vector so far
(function BodyToVecOperandHelper (i64 i64 Body VecOperandBase) VecOperand :unextractable)

(rewrite (BodyToVecOperand body-len body)
         (BodyToVecOperandHelper 0 body-len body (vec-of)) :ruleset subst)
(rule
  ((= helper (BodyToVecOperandHelper index body-len body so-far))
   (< index body-len))
  ((union helper
          (BodyToVecOperandHelper (+ index 1) body-len body
                                  (vec-push so-far 
                                   (Project index body)))))
  :ruleset subst)

(rule
  ((= helper (BodyToVecOperandHelper index body-len body so-far))
   (= index body-len))
  ((union helper (VO so-far)))
  :ruleset subst)

;; Converts (vec n1 n2 ...) to (vec (Arg n1) (Arg n2) ...)
(function MapArgOverVec (VecInt) VecOperand :unextractable)
(function MapArgOverVecHelper (VecIntBase i64 VecOperandBase) VecOperand :unextractable)
(let vec-operand-empty (vec-pop (vec-of (Arg 0))))
(rewrite (MapArgOverVec (VI vec)) 
         (MapArgOverVecHelper vec 0 vec-operand-empty)
      :ruleset subst)
(rewrite (MapArgOverVecHelper vec i curr)
         (MapArgOverVecHelper vec (+ i 1) (vec-push curr (Arg (vec-get vec i))))
      :when ((< i (vec-length vec)))
      :ruleset subst)
(rewrite (MapArgOverVecHelper vec i curr)
         (VO curr)
      :when ((= i (vec-length vec)))
      :ruleset subst)

(function VecAppend (VecOperand VecOperand) VecOperand)
(rewrite (VecAppend (VO v1) (VO v2)) 
         (VO (vec-append v1 v2))
      :ruleset subst)

;; Converts (vec n1 n2 ...) to (map-of n1 (Arg offset) n2 (Arg (+ offset 1)) ...)
;;                           vec    offset
(function BuildSubstMapFrom (VecInt i64) MapIntOperand :unextractable)
(function BuildSubstMapFromHelper (VecIntBase i64 i64 MapIntOperandBase) MapIntOperand :unextractable)
(rewrite (BuildSubstMapFrom (VI vi) offset) 
         ;; because type checking for actions is broken, we 
         ;; have to make sure there is only one Map sort right now
         (BuildSubstMapFromHelper vi offset 0 (map-empty))
         :ruleset subst)
(rewrite (BuildSubstMapFromHelper vi offset i m)
         (BuildSubstMapFromHelper vi offset (+ i 1) (map-insert m (vec-get vi i) (Arg (+ i offset))))
      :when ((< i (vec-length vi)))
      :ruleset subst)
(rewrite (BuildSubstMapFromHelper vi offset i m)
         (MIO m)
      :when ((= i (vec-length vi)))
      :ruleset subst)
