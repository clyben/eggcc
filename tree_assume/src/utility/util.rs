#[cfg(test)]
use crate::{egglog_test, interpreter::Value};

#[test]
fn test_list_util() -> crate::Result {
    let build = "
		(let list (Cons (Const (Int 0))
                  (Cons (Const (Int 1))
                  (Cons (Const (Int 2))
                  (Cons (Const (Int 3))
                  (Cons (Const (Int 4)) (Nil)))))))
		(let expr (Switch (Const (Int 1)) list))
	";
    let check = "
		(check (= (ListExpr-ith list 1) (Const (Int 1))))
        (check (= (ListExpr-ith list 4) (Const (Int 4))))
        (check (= (ListExpr-length list) 5))
	";
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
    )
}

#[test]
fn append_test() -> crate::Result {
    let build = "
        (let appended
            (Append
                (Cons (Const (Int 0)) (Cons (Const (Int 1)) (Nil)))
                (Const (Int 2))))
    ";

    let check = "
        (check (
            =
            (Cons (Const (Int 0)) (Cons (Const (Int 1)) (Cons (Const (Int 2)) (Nil))))
            appended
        ))
    ";
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
    )
}