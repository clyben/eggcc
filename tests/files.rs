use eggcc::util::Run;
use insta::assert_snapshot;
use libtest_mimic::Trial;

fn generate_tests(glob: &str) -> Vec<Trial> {
    let mut trials = vec![];
    let mut mk_trial = |run: Run| {
        trials.push(Trial::test(run.name(), move || {
            let result = run.run();

            if let Some(interpreted) = result.result_interpreted {
                assert_eq!(result.original_interpreted, interpreted);
            } else {
                assert_snapshot!(run.name(), result.visualization);
            }
            Ok(())
        }))
    };

    for entry in glob::glob(glob).unwrap() {
        let f = entry.unwrap();
        let name = f
            .file_stem()
            .unwrap()
            .to_string_lossy()
            .replace(['.', '-', ' '], "_");

        // TODO optimizer doesn't support these yet
        let banned = ["queens_func", "unstructured", "implicit_return", "fib"];
        if banned.iter().any(|b| name.contains(b)) || f.to_str().unwrap().contains("failing") {
            continue;
        }

        for run in Run::all_configurations_for(f) {
            mk_trial(run);
        }
    }

    trials
}

fn main() {
    let args = libtest_mimic::Arguments::from_args();
    let tests = generate_tests("tests/**/*.bril");
    libtest_mimic::run(&args, tests).exit();
}
