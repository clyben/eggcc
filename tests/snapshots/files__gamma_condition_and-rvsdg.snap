---
source: tests/files.rs
expression: visualization.result
---
<svg
	height="1050"
	version="1.1"
	width="1475"
	xmlns="http://www.w3.org/2000/svg">
	<g
		transform="translate(0, 0)">
		<rect
			fill="white"
			height="1050"
			rx="10"
			stroke="black"
			stroke-width="2"
			width="1475">
		</rect>
		<g>
			<path
				d="M 491.66666 0V 10Q 491.66666 20 481.66666 20H 393.33334Q 383.33334 20 383.33334 30V 200"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 491.66666 0V 10Q 491.66666 20 501.66666 20H 1106.6666Q 1116.6666 20 1116.6666 30V 500"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 983.3333 0V 20Q 983.3333 30 973.3333 30H 968.3333Q 958.3333 30 958.3333 40V 500"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 100 150V 170Q 100 180 110 180H 406.66666Q 416.66666 180 416.66666 190V 200"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 100 150V 170Q 100 180 110 180H 523.3333Q 533.3333 180 533.3333 190V 200"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 100 150V 170Q 100 180 110 180H 1185.8333Q 1195.8333 180 1195.8333 190V 500"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 250 150V 160Q 250 170 260 170H 556.6667Q 566.6667 170 566.6667 180V 200"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 250 150V 160Q 250 170 260 170H 1027.5Q 1037.5 170 1037.5 180V 500"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 400 300V 320Q 400 330 410 330H 673.3333Q 683.3333 330 683.3333 340V 350"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 550 300V 310Q 550 320 560 320H 706.6667Q 716.6667 320 716.6667 330V 350"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 700 450V 465Q 700 475 710 475H 869.1667Q 879.1667 475 879.1667 485V 500"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 958.3333 850V 860Q 958.3333 870 968.3333 870H 1381.6666Q 1391.6666 870 1391.6666 880V 900"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 1116.6666 850V 870Q 1116.6666 880 1126.6666 880H 1348.3334Q 1358.3334 880 1358.3334 890V 900"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
			<path
				d="M 1375 1000V 1015Q 1375 1025 1365 1025H 747.5Q 737.5 1025 737.5 1035V 1050"
				fill="transparent"
				stroke="black"
				stroke-linecap="round"
				stroke-width="2">
			</path>
		</g>
		<g>
			<g
				transform="translate(50, 50)">
				<g>
					<rect
						fill="#FFFF80"
						height="100"
						rx="10"
						stroke="black"
						stroke-width="2"
						width="100">
					</rect>
					<text
						fill="black"
						font-size="50"
						text-anchor="middle"
						x="50"
						y="62.5">
						2
					</text>
				</g>
				<circle
					cx="50"
					cy="100"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
			</g>
			<g
				transform="translate(200, 50)">
				<g>
					<rect
						fill="#FFFF80"
						height="100"
						rx="10"
						stroke="black"
						stroke-width="2"
						width="100">
					</rect>
					<text
						fill="black"
						font-size="50"
						text-anchor="middle"
						x="50"
						y="62.5">
						3
					</text>
				</g>
				<circle
					cx="50"
					cy="100"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
			</g>
			<g
				transform="translate(350, 200)">
				<g>
					<rect
						fill="#FFFF80"
						height="100"
						rx="10"
						stroke="black"
						stroke-width="2"
						width="100">
					</rect>
					<text
						fill="black"
						font-size="50"
						text-anchor="middle"
						x="50"
						y="62.5">
						lt
					</text>
				</g>
				<circle
					cx="33.333332"
					cy="0"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="66.666664"
					cy="0"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="50"
					cy="100"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
			</g>
			<g
				transform="translate(500, 200)">
				<g>
					<rect
						fill="#FFFF80"
						height="100"
						rx="10"
						stroke="black"
						stroke-width="2"
						width="100">
					</rect>
					<text
						fill="black"
						font-size="50"
						text-anchor="middle"
						x="50"
						y="62.5">
						lt
					</text>
				</g>
				<circle
					cx="33.333332"
					cy="0"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="66.666664"
					cy="0"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="50"
					cy="100"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
			</g>
			<g
				transform="translate(650, 350)">
				<g>
					<rect
						fill="#FFFF80"
						height="100"
						rx="10"
						stroke="black"
						stroke-width="2"
						width="100">
					</rect>
					<text
						fill="black"
						font-size="50"
						text-anchor="middle"
						x="50"
						y="62.5">
						and
					</text>
				</g>
				<circle
					cx="33.333332"
					cy="0"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="66.666664"
					cy="0"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="50"
					cy="100"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
			</g>
			<g
				transform="translate(800, 500)">
				<rect
					fill="#80FF80"
					height="350"
					rx="10"
					stroke="black"
					stroke-width="2"
					width="475">
				</rect>
				<text
					fill="black"
					font-size="50"
					x="25"
					y="50">
					match
				</text>
				<g>
					<g
						transform="translate(25, 125)">
						<rect
							fill="white"
							height="200"
							rx="10"
							stroke="black"
							stroke-width="2"
							width="200">
						</rect>
						<g>
							<path
								d="M 40 0V 6Q 40 16 50 16H 106.666664Q 116.666664 16 116.666664 26V 50"
								fill="transparent"
								stroke="black"
								stroke-linecap="round"
								stroke-width="2">
							</path>
							<path
								d="M 80 0V 170Q 80 180 90 180H 123.33333Q 133.33333 180 133.33333 190V 200"
								fill="transparent"
								stroke="black"
								stroke-linecap="round"
								stroke-width="2">
							</path>
							<path
								d="M 160 0V 24Q 160 34 150 34H 93.33333Q 83.33333 34 83.33333 44V 50"
								fill="transparent"
								stroke="black"
								stroke-linecap="round"
								stroke-width="2">
							</path>
							<path
								d="M 100 150V 160Q 100 170 90 170H 76.666664Q 66.666664 170 66.666664 180V 200"
								fill="transparent"
								stroke="black"
								stroke-linecap="round"
								stroke-width="2">
							</path>
						</g>
						<g>
							<g
								transform="translate(50, 50)">
								<g>
									<rect
										fill="#FFFF80"
										height="100"
										rx="10"
										stroke="black"
										stroke-width="2"
										width="100">
									</rect>
									<text
										fill="black"
										font-size="50"
										text-anchor="middle"
										x="50"
										y="62.5">
										PRINT
									</text>
								</g>
								<circle
									cx="33.333332"
									cy="0"
									fill="white"
									r="4"
									stroke="black"
									stroke-width="2">
								</circle>
								<circle
									cx="66.666664"
									cy="0"
									fill="white"
									r="4"
									stroke="black"
									stroke-width="2">
								</circle>
								<circle
									cx="50"
									cy="100"
									fill="white"
									r="4"
									stroke="black"
									stroke-width="2">
								</circle>
							</g>
						</g>
						<g>
							<circle
								cx="40"
								cy="0"
								fill="white"
								r="4"
								stroke="black"
								stroke-width="2">
							</circle>
							<circle
								cx="80"
								cy="0"
								fill="white"
								r="4"
								stroke="black"
								stroke-width="2">
							</circle>
							<circle
								cx="120"
								cy="0"
								fill="white"
								r="4"
								stroke="black"
								stroke-width="2">
							</circle>
							<circle
								cx="160"
								cy="0"
								fill="white"
								r="4"
								stroke="black"
								stroke-width="2">
							</circle>
						</g>
						<g>
							<circle
								cx="66.666664"
								cy="200"
								fill="white"
								r="4"
								stroke="black"
								stroke-width="2">
							</circle>
							<circle
								cx="133.33333"
								cy="200"
								fill="white"
								r="4"
								stroke="black"
								stroke-width="2">
							</circle>
						</g>
					</g>
					<text
						fill="black"
						font-size="50"
						text-anchor="middle"
						x="125"
						y="100">
						0
					</text>
				</g>
				<g>
					<g
						transform="translate(250, 125)">
						<rect
							fill="white"
							height="200"
							rx="10"
							stroke="black"
							stroke-width="2"
							width="200">
						</rect>
						<g>
							<path
								d="M 40 0V 6Q 40 16 50 16H 106.666664Q 116.666664 16 116.666664 26V 50"
								fill="transparent"
								stroke="black"
								stroke-linecap="round"
								stroke-width="2">
							</path>
							<path
								d="M 80 0V 170Q 80 180 90 180H 123.33333Q 133.33333 180 133.33333 190V 200"
								fill="transparent"
								stroke="black"
								stroke-linecap="round"
								stroke-width="2">
							</path>
							<path
								d="M 120 0V 18Q 120 28 110 28H 93.33333Q 83.33333 28 83.33333 38V 50"
								fill="transparent"
								stroke="black"
								stroke-linecap="round"
								stroke-width="2">
							</path>
							<path
								d="M 100 150V 160Q 100 170 90 170H 76.666664Q 66.666664 170 66.666664 180V 200"
								fill="transparent"
								stroke="black"
								stroke-linecap="round"
								stroke-width="2">
							</path>
						</g>
						<g>
							<g
								transform="translate(50, 50)">
								<g>
									<rect
										fill="#FFFF80"
										height="100"
										rx="10"
										stroke="black"
										stroke-width="2"
										width="100">
									</rect>
									<text
										fill="black"
										font-size="50"
										text-anchor="middle"
										x="50"
										y="62.5">
										PRINT
									</text>
								</g>
								<circle
									cx="33.333332"
									cy="0"
									fill="white"
									r="4"
									stroke="black"
									stroke-width="2">
								</circle>
								<circle
									cx="66.666664"
									cy="0"
									fill="white"
									r="4"
									stroke="black"
									stroke-width="2">
								</circle>
								<circle
									cx="50"
									cy="100"
									fill="white"
									r="4"
									stroke="black"
									stroke-width="2">
								</circle>
							</g>
						</g>
						<g>
							<circle
								cx="40"
								cy="0"
								fill="white"
								r="4"
								stroke="black"
								stroke-width="2">
							</circle>
							<circle
								cx="80"
								cy="0"
								fill="white"
								r="4"
								stroke="black"
								stroke-width="2">
							</circle>
							<circle
								cx="120"
								cy="0"
								fill="white"
								r="4"
								stroke="black"
								stroke-width="2">
							</circle>
							<circle
								cx="160"
								cy="0"
								fill="white"
								r="4"
								stroke="black"
								stroke-width="2">
							</circle>
						</g>
						<g>
							<circle
								cx="66.666664"
								cy="200"
								fill="white"
								r="4"
								stroke="black"
								stroke-width="2">
							</circle>
							<circle
								cx="133.33333"
								cy="200"
								fill="white"
								r="4"
								stroke="black"
								stroke-width="2">
							</circle>
						</g>
					</g>
					<text
						fill="black"
						font-size="50"
						text-anchor="middle"
						x="350"
						y="100">
						1
					</text>
				</g>
				<circle
					cx="79.166664"
					cy="0"
					fill="black"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="158.33333"
					cy="0"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="237.5"
					cy="0"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="316.66666"
					cy="0"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="395.8333"
					cy="0"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="158.33333"
					cy="350"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="316.66666"
					cy="350"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
			</g>
			<g
				transform="translate(1325, 900)">
				<g>
					<rect
						fill="#FFFF80"
						height="100"
						rx="10"
						stroke="black"
						stroke-width="2"
						width="100">
					</rect>
					<text
						fill="black"
						font-size="50"
						text-anchor="middle"
						x="50"
						y="62.5">
						PRINT
					</text>
				</g>
				<circle
					cx="33.333332"
					cy="0"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="66.666664"
					cy="0"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
				<circle
					cx="50"
					cy="100"
					fill="white"
					r="4"
					stroke="black"
					stroke-width="2">
				</circle>
			</g>
		</g>
		<g>
			<circle
				cx="491.66666"
				cy="0"
				fill="white"
				r="4"
				stroke="black"
				stroke-width="2">
			</circle>
			<circle
				cx="983.3333"
				cy="0"
				fill="white"
				r="4"
				stroke="black"
				stroke-width="2">
			</circle>
		</g>
		<g>
			<circle
				cx="737.5"
				cy="1050"
				fill="white"
				r="4"
				stroke="black"
				stroke-width="2">
			</circle>
		</g>
	</g>
</svg>

